Source/controls/plrctrls.cpp

```diff
@@ -1409,37 +1408,75 @@ bool IsStickMovementSignificant()
 	    || rightStickX != 0 || rightStickY != 0;
 }
 
-ControlTypes GetInputTypeFromEvent(const SDL_Event &event)
+ControlTypeWithGamepad GetInputTypeFromEvent(const SDL_Event &event)
 {
-	if (IsAnyOf(event.type, SDL_KEYDOWN, SDL_KEYUP))
-		return ControlTypes::KeyboardAndMouse;
 #ifdef USE_SDL1
-	if (IsAnyOf(event.type, SDL_MOUSEBUTTONDOWN, SDL_MOUSEBUTTONUP, SDL_MOUSEMOTION))
-		return ControlTypes::KeyboardAndMouse;
+	if (IsAnyOf(event.type, SDL_MOUSEBUTTONDOWN, SDL_MOUSEBUTTONUP, SDL_MOUSEMOTION)) {
+		return {
+			.controlTypeCoarse = ControlTypes::KeyboardAndMouse,
+			.gamepadLayout = GamepadLayout::Generic
+		};
+	}
 #else
-	if (IsAnyOf(event.type, SDL_MOUSEBUTTONDOWN, SDL_MOUSEBUTTONUP))
-		return event.button.which == SDL_TOUCH_MOUSEID ? ControlTypes::VirtualGamepad : ControlTypes::KeyboardAndMouse;
-	if (event.type == SDL_MOUSEMOTION)
-		return event.motion.which == SDL_TOUCH_MOUSEID ? ControlTypes::VirtualGamepad : ControlTypes::KeyboardAndMouse;
-	if (event.type == SDL_MOUSEWHEEL)
-		return event.wheel.which == SDL_TOUCH_MOUSEID ? ControlTypes::VirtualGamepad : ControlTypes::KeyboardAndMouse;
-	if (IsAnyOf(event.type, SDL_FINGERDOWN, SDL_FINGERUP, SDL_FINGERMOTION))
-		return ControlTypes::VirtualGamepad;
+	GamepadLayout gamepad = GameController::getLayout(event);
+
+	if (gamepad > GamepadLayout::Generic) {
+		return {
+			.controlTypeCoarse = gamepad < GamepadLayout::Virtual ? ControlTypes::Gamepad : ControlTypes::VirtualGamepad,

```

Comment:
By the way, treating an enum as if it was an integer and assuming knolage of the underlying type will also fail when using scoped enums.
```suggestion
			.controlTypeCoarse = gamepad != GamepadLayout::Virtual ? ControlTypes::Gamepad : ControlTypes::VirtualGamepad,
```

If you mean to apply this to more then 1 option then use `IsAnyOf(gamepad, A, B, C)` or `IsNoneOf(gamepad, A, B, C)`